### ISSUE TITLE
Assertion `LoopCost && "Non-zero loop cost expected"' failed

### ISSUE THREAD

**Issue by nikic** [llvm/llvm-project#54413]
The following input asserts under `opt -passes=loop(indvars),loop-vectorize`:
```
target triple = "x86_64-unknown-linux-gnu"

define void @test(i64* %ptr.base) {
entry:
  br label %loop1

loop1:
  br i1 true, label %loop1.latch, label %loop2.preheader

loop2.preheader:
  br label %loop2

loop2:
  %iv = phi i64 [ 0, %loop2.preheader ], [ %iv.next, %loop2 ]
  %ptr = phi i64* [ %ptr.base, %loop2.preheader ], [ %ptr.next, %loop2 ]
  %iv.next = add nuw nsw i64 %iv, 1
  %ptr.next = getelementptr inbounds i64, i64* %ptr, i64 1
  %cmp = icmp eq i64 %iv, 1024
  br i1 %cmp, label %loop2.exit, label %loop2

loop2.exit:
  %ptr.next.lcssa = phi i64* [ %ptr.next, %loop2 ]
  br label %loop1.latch

loop1.latch:
  br label %loop1
}
```

```
opt: /home/npopov/repos/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:5923: unsigned int llvm::LoopVectorizationCostModel::selectInterleaveCount(llvm::ElementCount, unsigned int): Assertion `LoopCost && "Non-zero loop cost expected"' failed.
[...]
 #8 0x000000000341d17c llvm::LoopVectorizationCostModel::selectInterleaveCount(llvm::ElementCount, unsigned int) (build/bin/opt+0x341d17c)
 #9 0x0000000003424d84 llvm::LoopVectorizePass::processLoop(llvm::Loop*) (build/bin/opt+0x3424d84)
#10 0x000000000342738e llvm::LoopVectorizePass::runImpl(llvm::Function&, llvm::ScalarEvolution&, llvm::LoopInfo&, llvm::TargetTransformInfo&, llvm::DominatorTree&, llvm::BlockFrequencyInfo&, llvm::TargetLibraryInfo*, llvm::DemandedBits&, llvm::AAResults&, llvm::AssumptionCache&, std::function<llvm::LoopAccessInfo const& (llvm::Loop&)>&, llvm::OptimizationRemarkEmitter&, llvm::ProfileSummaryInfo*) (build/bin/opt+0x342738e)
#11 0x0000000003427bbc llvm::LoopVectorizePass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build/bin/opt+0x3427bbc)
```

**Comment by nikic** [comment]
For reference, this is how it looks like after `indvars`:
```
define void @test(i64* %ptr.base) {
entry:
  br label %loop1

loop1:                                            ; preds = %loop1.latch, %entry
  br i1 true, label %loop1.latch, label %loop2.preheader

loop2.preheader:                                  ; preds = %loop1
  br label %loop2

loop2:                                            ; preds = %loop2, %loop2.preheader
  %ptr = phi i64* [ %ptr.base, %loop2.preheader ], [ %ptr.next, %loop2 ]
  %ptr.next = getelementptr inbounds i64, i64* %ptr, i64 1
  br i1 true, label %loop2.exit, label %loop2

loop2.exit:                                       ; preds = %loop2
  %ptr.next.lcssa = phi i64* [ %ptr.next, %loop2 ]
  br label %loop1.latch

loop1.latch:                                      ; preds = %loop2.exit, %loop1
  br label %loop1
}
```

The reason why we need the indvars run is to keep the cached SCEV BECount, which is originally 1025 and becomes 1 after the transform. As the loop is in dead code, retaining the old BECount is legal.

This leaves us with a loop that is both zero-cost according to the cost model, *and* has a non-zero/one BECount, so we do try to vectorize it.

I assume the fix here would be to just convert that assertion into a check?

cc @fhahn

**Comment by fhahn** [comment]
> I assume the fix here would be to just convert that assertion into a check?

The assertion looks overly aggressive. But I think it would be better to just compute the cost earlier with user-specified VFs. Let me take a look.

**Comment by fhahn** [comment]
> > I assume the fix here would be to just convert that assertion into a check?
> 
> The assertion looks overly aggressive. But I think it would be better to just compute the cost earlier with user-specified VFs. Let me take a look.

Sorry for the delay, I just landed a fix. Computing the cost earlier turned out unfeasible, as. there are multiple code paths that would need updating and it would unnecessarily complicate earlier logic.

**Comment by tstellar** [comment]
@fhahn Is this safe to backport? https://github.com/llvm/llvm-project/commit/ecb4171dcbf1b433c9963fd605a74898303e850d

**Comment by fhahn** [comment]
Yes that should be safe, thanks!

**Comment by tstellar** [comment]
Merged: fe8a27acd716
